File: Android - Native/discover_destination/app/src/main/java/com/example/discover_destination/screens/AddScreen.kt
package com.example.discover_destination.screens

import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.example.discover_destination.domain.Destination
import com.example.discover_destination.viewmodel.DestinationsViewModel

@Composable
fun AddScreen(
    viewModel: DestinationsViewModel = hiltViewModel(),
    onFinishedOperation: () -> Unit,
    user: String
) {
    val cityFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val countryFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val tourismFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val imageFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val descriptionFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val attractionFieldValue = remember { mutableStateOf(TextFieldValue()) }
    val attractionsValue: MutableState<String> = remember { mutableStateOf("") }
    val errorValue = remember { mutableStateOf(TextFieldValue()) }

    Column(
        horizontalAlignment = Alignment.End,
        modifier = Modifier.padding(10.dp)
    ) {
        Text(
                text = "Add Destination",
                fontSize = 35.sp
            )
        Column(
            horizontalAlignment = Alignment.Start
        ) {
            Text(
                text = "City",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = cityFieldValue.value,
                onValueChange = { cityFieldValue.value = it }
            )
            Text(
                text = "Country",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = countryFieldValue.value,
                onValueChange = { countryFieldValue.value = it }
            )
            Text(
                text = "Type of Tourism",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = tourismFieldValue.value,
                onValueChange = { tourismFieldValue.value = it }
            )
            Text(
                text = "Image URL",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = imageFieldValue.value,
                onValueChange = { imageFieldValue.value = it }
            )
            Text(
                text = "Description",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = descriptionFieldValue.value,
                onValueChange = { descriptionFieldValue.value = it }
            )
            Text(
                text = "Tourist Attraction",
                fontSize = 15.sp
            )
            Row(){
                OutlinedTextField(
                    value = attractionFieldValue.value,
                    onValueChange = { attractionFieldValue.value = it }
                )
                OutlinedButton(onClick = {
                    attractionsValue.value += attractionFieldValue.value.text + ", "
                    attractionFieldValue.value = TextFieldValue() },
                    modifier = Modifier.size(55.dp)
                ) {
                    Icon(
                        Icons.Filled.Add,
                        contentDescription = null
                    )
                }
            }
            Text(
                text = "List of Tourist Attractions:",
                fontSize = 15.sp
            )
            Text(
                text = attractionsValue.value,
                fontSize = 20.sp
            )
            OutlinedButton(onClick = {
                val destination = Destination(cityFieldValue.value.text, countryFieldValue.value.text,
                    tourismFieldValue.value.text, descriptionFieldValue.value.text,
                    imageFieldValue.value.text, attractionsValue.value.split(", ").dropLast(1), user)
                if (validate(destination)){
                    viewModel.add(destination)
                    onFinishedOperation()
                }
                else {
                    errorValue.value = TextFieldValue("Invalid data!")
                }
            },
                modifier = Modifier.size(400.dp, 55.dp)
            ) {
                Text(
                    text = "ADD DESTINATION",
                    fontSize = 20.sp,
                    color = Color.Black
                )
            }
            Text(
                text = errorValue.value.text,
                fontSize = 15.sp,
                color = Color.Red
            )
        }
    }
}

fun validate(destination: Destination): Boolean{
    return destination.city.length > 0 && destination.country.length > 0 &&
            destination.description.length > 0 && destination.image.length > 0 &&
            destination.tourism.length > 0 && !destination.touristAttractions.isEmpty()
}

File: Flutter - Non-Native/discover_destination/lib/view/add_screen.dart
import 'package:discover_destination/model/destination.dart';
import 'package:discover_destination/view/start_screen.dart';
import 'package:discover_destination/viewmodel/destinations_view_model.dart';
import 'package:flutter/material.dart';

class AddScreen extends StatefulWidget {
  DestinationsViewModel destinationsViewModel;
  String user;

  AddScreen(this.destinationsViewModel, this.user, {Key? key})
      : super(key: key);

  @override
  _AddState createState() => _AddState(destinationsViewModel, user);
}

class _AddState extends State<AddScreen> {
  DestinationsViewModel destinationsViewModel;
  String user;
  final cityTextFieldController = TextEditingController();
  final countryTextFieldController = TextEditingController();
  final tourismTextFieldController = TextEditingController();
  final imageTextFieldController = TextEditingController();
  final descriptionTextFieldController = TextEditingController();
  final attractionTextFieldController = TextEditingController();
  final attractionsTextFieldController = TextEditingController();

  _AddState(this.destinationsViewModel, this.user);

  bool validate(Destination destination) {
    return destination.city.length > 0 &&
        destination.country.length > 0 &&
        destination.description.length > 0 &&
        destination.image.length > 0 &&
        destination.tourism.length > 0 &&
        !destination.touristAttractions.isEmpty;
  }

  @override
  Widget build(BuildContext context) {
    Destination destination;
    List<String> attractions;

    return Scaffold(
        backgroundColor: Colors.white,
        body: SingleChildScrollView(
            child: Column(children: <Widget>[
          Padding(
            padding: const EdgeInsets.only(top: 60.0),
            child: Center(
              child: Container(
                  height: 70,
                  child: Text("Add Destination",
                      textAlign: TextAlign.left,
                      style: TextStyle(fontSize: 45.0))),
            ),
          ),
          Padding(
              padding: EdgeInsets.symmetric(horizontal: 15),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  TextField(
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'City',
                    ),
                    controller: cityTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  TextField(
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Country',
                    ),
                    controller: countryTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  TextField(
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Type of Tourism',
                    ),
                    controller: tourismTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  TextField(
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Image URL',
                    ),
                    controller: imageTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  TextField(
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Description',
                    ),
                    controller: descriptionTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        flex: 85,
                        child: TextField(
                          decoration: InputDecoration(
                            border: OutlineInputBorder(),
                            labelText: 'Tourist Attraction',
                          ),
                          controller: attractionTextFieldController,
                        ),
                      ),
                      Expanded(
                        flex: 15,
                        child: FlatButton(
                            onPressed: () => {
                                  attractionsTextFieldController.text +=
                                      attractionTextFieldController.text + ", ",
                                  attractionTextFieldController.text = ""
                                },
                            child: Icon(Icons.add),
                            ),
                      )
                    ],
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  Text("List of Tourist Attractions:",
                      textAlign: TextAlign.left,
                      style: TextStyle(fontSize: 20.0)),
                  TextFormField(
                    readOnly: true,
                    decoration: InputDecoration(border: OutlineInputBorder()),
                    controller: attractionsTextFieldController,
                  ),
                  SizedBox(
                    height: 10.0,
                  ),
                  Center(
                    child:
                  OutlinedButton(
                    onPressed: () => {
                      attractions =
                          attractionsTextFieldController.text.split(", "),
                      destination = Destination(
                          cityTextFieldController.text,
                          countryTextFieldController.text,
                          tourismTextFieldController.text,
                          descriptionTextFieldController.text,
                          imageTextFieldController.text,
                          attractions.sublist(0, attractions.length - 1),
                          user),
                      if (validate(destination))
                        {
                          destinationsViewModel.add(destination),
                          Navigator.push(
                              context,
                              MaterialPageRoute(
                                  builder: (_) =>
                                      StartScreen(destinationsViewModel, user)))
                        }
                      else
                        {
                          showDialog(
                              context: context,
                              builder: (ctx) => AlertDialog(
                                    title: Text("Invalid data"),
                                    content:
                                        Text("All fields must be completed!"),
                                    actions: <Widget>[
                                      FlatButton(
                                        onPressed: () {
                                          Navigator.of(ctx).pop();
                                        },
                                        child: Text("OK"),
                                      ),
                                    ],
                                  ))
                        }
                    },
                    child: Text(
                      'ADD DESTINATION',
                      style: TextStyle(color: Colors.black, fontSize: 25),
                    ),
                  ))
                ],
              ))
        ])));
  }
}
